net/dnsfallback/dnsfallback.go

```diff
@@ -44,76 +45,140 @@ var (
-				Dialer: netns.NewDialer(logf, netMon),
-				Logf:   logf,
-			}
-			addrs, minTTL, err := resolver.Resolve(ctx, host)
-			if err != nil {
-				logf("error using recursive resolver: %v", err)
-				metricRecursiveErrors.Add(1)
-				return
-			}
+	// for tests
+	waitForCompare bool
+}
 
-			compareAddr := func(a, b netip.Addr) int { return a.Compare(b) }
-			slices.SortFunc(addrs, compareAddr)
+func (fr *fallbackResolver) Lookup(ctx context.Context, host string) ([]netip.Addr, error) {
+	// If they've explicitly disabled the recursive resolver with the legacy
+	// TS_DNSFALLBACK_DISABLE_RECURSIVE_RESOLVER envknob or not set the
+	// newer TS_DNSFALLBACK_RECURSIVE_RESOLVER to true, then don't use the
+	// recursive resolver. (tailscale/corp#15261) In the future, we might
+	// change the default (the opt.Bool being unset) to mean enabled.
+	if disableRecursiveResolver() || !optRecursiveResolver().EqualBool(true) {
+		return lookup(ctx, host, fr.logf, fr.netMon)
+	}
 
-			// Wait for a response from the main function
-			oldAddrs := <-addrsCh
-			slices.SortFunc(oldAddrs, compareAddr)
+	addrsCh := make(chan []netip.Addr, 1)
 
-			matches := slices.Equal(addrs, oldAddrs)
+	// Run the recursive resolver in the background so we can
+	// compare the results. For tests, we also allow waiting for the
+	// comparison to complete; normally, we do this entirely asynchronously
+	// so as not to block the caller.
+	var done chan struct{}
+	if fr.waitForCompare {
+		done = make(chan struct{})
+		go func() {
+			defer close(done)
+			fr.compareWithRecursive(ctx, addrsCh, host)
+		}()
+	} else {
+		go fr.compareWithRecursive(ctx, addrsCh, host)
+	}
 
-			logf("bootstrap DNS comparison: matches=%v oldAddrs=%v addrs=%v minTTL=%v", matches, oldAddrs, addrs, minTTL)
+	addrs, err := lookup(ctx, host, fr.logf, fr.netMon)
+	if err != nil {
+		addrsCh <- nil
+		return nil, err
+	}
 
-			if matches {
-				metricRecursiveMatches.Add(1)
-			} else {
-				metricRecursiveMismatches.Add(1)
-			}
-		}()
+	addrsCh <- slices.Clone(addrs)
+	if fr.waitForCompare {
+		<-done
+	}
+	return addrs, nil
+}
 
-		addrs, err := lookup(ctx, host, logf, netMon)
+func (fr *fallbackResolver) compareWithRecursive(
+	ctx context.Context,
+	addrsCh <-chan []netip.Addr,
+	host string,
+) {
+	logf := logger.WithPrefix(fr.logf, "recursive: ")
+
+	// Ensure that we catch panics while we're testing this
+	// code path; this should never panic, but we don't
+	// want to take down the process by having the panic
+	// propagate to the top of the goroutine's stack and
+	// then terminate.
+	defer func() {
+		if r := recover(); r != nil {
+			logf("bootstrap DNS: recovered panic: %v", r)
+			metricRecursiveErrors.Add(1)
+		}
+	}()
+
+	// Don't resolve the same host multiple times
+	// concurrently; if we end up in a tight loop, this can
+	// take up a lot of CPU.
+	var didRun bool
+	result, err, _ := fr.sf.Do(host, func() (resolveResult, error) {
+		didRun = true
+		resolver := recursive.Resolver{
```

Comment:
methods are on pointer so do `&recursive.Resolver` here just for clarity (and to avoid accidental value copies of `resolver`, which isn't a problem here, but in general)